### ISSUE TITLE
Introduce 'answer_with" (renaming answer_as_code)? Add answer_with_sql? Rename add_tools?

### ISSUE THREAD

**Issue by lukakoning** [tjarkvandemerwe/tidyprompt#64]
Currently we have answer_as_code which is a hybrid between answer_as and a tool 

We could introduce answer_with which could be a more appropriate name, this class of prompt wraps could
hold all prompt wraps which may serve as 'answer_as' but may also be used as tools (where the LLM can make a call,
and the call's results may be used for formulating a final answer)

We could, e.g., also introduce answer_with_sql where the LLM gets passed a database object and may generate SQL, which is optionally evaluated for use as a tool, or optionally evaluated to go straight from prompt to SQL result (data)

Maybe rename add_tools to answer_using. So: 
* answer_as = format specifically, 
* answer_with = format (typically code), optionally a result of the code, or code is used by LLM as a tool; 
* answer_using = exclusively a tool

**Comment by tjarkvandemerwe** [comment]
I like the idea of making a distinction between the 'return code' and 'use code to return object' functions. The latter being tool calling.

However I do not yet fully understand when to use the answer_with hybrid function in this spectrum. What would we lack if we use the following interface:

- answer_as_r()
- answer_as_sql()
- answer_as_python() 

etc. returning code. Extracting code from answer, check if at least and at most 1 code block is returned. Maybe optional arguments like: save_location. But no evaluation of code.

- answer_using_r()
- answer_using_sql()
- answer_using_python()

And have answer_using_<language/tool> have the arguments: functions, packages, (data_)objects, to add usable objects to the environment.
And have some kind of return mode. Default it uses the tool to find information to return a character object containing the answer. But you can also add different modes, to generate certain objects. With constraints on using only the tool and objects provided. Do we miss a use-case in this scenario?

**Comment by lukakoning** [comment]
answer_as/answer_using seems look a good idea, especially if we design it so that the answer_using can use the same functions the related answer_as is using.

for answer_using_... I think a distinction between answer_using_r (code) and answer_using_functions (currently add_tools) is likely needed, the former simply parses R code from the LLM response while the latter parses function calls in JSON form (optionally using native LLM provider configuration where they LLM provider will handle it). So they work quite differently and probably can't be integrated. Would be interesting though to also give function descriptions when functions are loaded as R object into answer_using_r (like the current answer_as_code does for, e.g., dataframes)

**Comment by lukakoning** [comment]
Om prompt wraps op elkaar te laten inspelen kan inderdaad wel tricky zijn (ook om het voor user transparant te houden wat er gebeurt?). Dus beetje mixed feelings, maar misschien kan het ook wel heel clean worden! Mogelijk is nodig dat modify/extraction/validation fns toegang krijgen tot de gehele list prompt wraps van de tidyprompt die wordt geëvalueerd?

**Comment by tjarkvandemerwe** [comment]
Yes, could indeed mean that results become unclean, however I would expect to be able to combine answer_using_r + answer_as_int and get an int back, but make it use r to get to the answer. And the same for answer_as_dataframe.

There are two options that I could think of:
- Conditional prompting in the prompt text, so, 'answer by returning a json frame, or, if other parts of the prompt enables you to use a programming language, generate the dataframe in the programming language', maybe something like this would work, but it is a bit tricky ad adds complexity aka confusion to the prompts.
- Look around wrappers, that indeed check the prompt stack and change their behaviour. But we might need a names attribute in the prompt wrapper object, to check for specific names in the stack.

**Comment by tjarkvandemerwe** [comment]
I think we could leverage S3 subclasses with this method, using tidyprompt_with_tool and the standard tidyprompt to differentiate function behaviors. By combining the order of operations with subclasses and S3, we can achieve flexible and adaptable functionality.

**Comment by lukakoning** [comment]
I added that in modify/extraction/validation/etc you can now access 'self', which is the tidyprompt object containing all prompt wraps. So this could also be used. 

Subclasses are worth exploring too then!

**Comment by lukakoning** [comment]
Closing for now, may reopen separate issues for some of the ideas mentioned here
